### PR TITLE
[SPARK-38075][SQL] Fix `hasNext` in `HiveScriptTransformationExec`'s process output iterator

### PR DESCRIPTION
**Pull request by bersprockets**
### What changes were proposed in this pull request?

Fix hasNext in HiveScriptTransformationExec's process output iterator to always return false if it had previously returned false.

### Why are the changes needed?

When hasNext on the process output iterator returns false, it leaves the iterator in a state (i.e., scriptOutputWritable is not null) such that the next call returns true.

The Guava Ordering used in TakeOrderedAndProjectExec will call hasNext on the process output iterator even after an earlier call had returned false. This results in fake rows when script transform is used with `order by` and `limit`. For example:

```
create or replace temp view t as
select * from values
(1),
(2),
(3)
as t(a);

select transform(a)
USING 'cat' AS (a int)
FROM t order by a limit 10;
```
This returns:
```
NULL
NULL
NULL
1
2
3
```

### Does this PR introduce _any_ user-facing change?

No, other than removing the correctness issue.

### How was this patch tested?

New unit test.

### CODE REVIEW

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/execution/HiveScriptTransformationExec.scala [hunk 1]
```diff
@@ -78,13 +78,17 @@ private[hive] case class HiveScriptTransformationExec(
...
+          return false
+        }
         try {
           if (scriptOutputWritable == null) {
             scriptOutputWritable = reusedWritableObject
 
             if (scriptOutputReader != null) {
               if (scriptOutputReader.next(scriptOutputWritable) <= 0) {
                 checkFailureAndPropagate(writerThread, null, proc, stderrBuffer)
+                completed = true
```
Comment (bersprockets) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/execution/HiveScriptTransformationExec.scala:91]: Alternatively, I could just set `scriptOutputWritable` to `null` here, and I wouldn't need the if statement at the top. That seemed to work. However, it feels a little unhygienic to read from an inputStream that has already returned EOF, so I added the flag.

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/HiveScriptTransformationSuite.scala [hunk 2]
```diff
@@ -621,4 +621,21 @@ class HiveScriptTransformationSuite extends BaseScriptTransformationSuite with T
       assert(e.contains("java.lang.ArithmeticException: long overflow"))
     }
   }
+
+  test("SPARK-38075: ORDER BY with LIMIT adds fake rows") {
```
Comment (dongjoon-hyun) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/HiveScriptTransformationSuite.scala:625]: `adds` -> `should not add`?

**Review by dongjoon-hyun — COMMENTED** [review]
Thank you, @bersprockets . Do you know when this bug started?

cc @viirya , too.

**Comment by bersprockets** [comment]
> Thank you, @bersprockets . Do you know when this bug started?

Not sure how far back, but at I can reproduce in 2.4.8, 3.1.2, 3.2.0, 3.2.1, and master.

**Review by dongjoon-hyun — COMMENTED** [review]
~It seems that `ORDER BY` clause issue and irrelevant to `LIMIT` clause. Could you confirm the following?~
```scala
scala> spark.version
res19: String = 3.2.1

scala> sql("SELECT transform(a) USING 'cat' AS (a int) FROM VALUES (1) t(a) ORDER BY a").show
+----+
|   a|
+----+
|null|
|   1|
+----+
```

**Comment by dongjoon-hyun** [comment]
Oh, never mind. I realized that the `spark-shell` add `limit` automatically. You're right. This is `ORDER BY .. LIMIT` issue.
```
spark-sql> SELECT version();
3.2.1 4f25b3f71238a00508a356591553f2dfa89f8290
Time taken: 0.208 seconds, Fetched 1 row(s)

spark-sql> SELECT transform(a) USING 'cat' AS (a int) FROM VALUES (1) t(a) ORDER BY a;
1
Time taken: 0.282 seconds, Fetched 1 row(s)
```

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/execution/HiveScriptTransformationExec.scala [hunk 3]
```diff
@@ -64,7 +64,7 @@ private[hive] case class HiveScriptTransformationExec(
       outputSoi: StructObjectInspector,
       hadoopConf: Configuration): Iterator[InternalRow] = {
     new Iterator[InternalRow] with HiveInspectors {
-      var curLine: String = null
```
Comment (dongjoon-hyun) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/execution/HiveScriptTransformationExec.scala:67]: Although this is irrelevant to this correctness issue, the clean-up looks okay.

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM.

**Review by viirya — APPROVED** [review]

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/execution/HiveScriptTransformationExec.scala [hunk 4]
```diff
@@ -64,7 +64,7 @@ private[hive] case class HiveScriptTransformationExec(
...
+      var completed = false
```
Comment (viirya) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/execution/HiveScriptTransformationExec.scala:67]: nit: private

**Comment by dongjoon-hyun** [comment]
Thank you for updates, @bersprockets . After compilation, I'll merge this.

**Comment by dongjoon-hyun** [comment]
Merged to master/3.2. For 3.1, there is a conflict.

**Comment by bersprockets** [comment]
> For 3.1, there is a conflict.
 
I will take a look at what needs to be done.

**Pull request closed without merging**